File: 309 Best Time to Buy and Sell Stock with Cooldown.c
"""
Created on Mon Nov. 09 11:49 2020

@author: Chiwen Chen
"""


int maxProfit(int* prices, int pricesSize){
  if (pricesSize < 2) return 0;
  int* p=(int*) malloc(sizeof(int) * pricesSize);
  p[0] = 0;
  int sell = -1;
  for (int i = 1; i < pricesSize; i++) {
      if (prices[i] <= prices[i-1]) {
          p[i] = p[i-1];
          continue;
      }
      
      int diff=prices[i] - prices[i-1];
      if (sell==i-1) {
          p[i]=p[i-1] + diff;
      } else if (sell>0){
          p[i] = p[i-3] + diff;// 0:2
          
          
          
          if (prices[i] > prices[sell]) {
              p[i] = fmax(p[i-1] + prices[i] - prices[sell], p[i]);
          } else if (sell == i-2 && p[i] < p[i-1]){
              p[i] = p[i-1];
              continue;
          }
      } 
      
      
      
      else {
          p[i] = diff;
      }
      sell = i;
  }
    
    
  int ret = p[pricesSize-1];
  free(p);//release malloc
  return ret;
}
//Runtime: 4 ms, faster than 66.67% of C online submissions for Best Time to Buy and Sell Stock with Cooldown.
//Memory Usage: 6.2 MB, less than 77.78% of C online submissions for Best Time to Buy and Sell Stock with Cooldown.

File: 309 Best Time to Buy and Sell Stock with Cooldown.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        
        if not prices or len(prices) < 2: return 0
        n = len(prices);
        buy, sell = [0] * n, [0] * n;
        buy[0] = -prices[0];
        buy[1] = max(-prices[0], -prices[1]);
        sell[1] = max(0, prices[1] - prices[0]);
        for i in range(2, n):
            buy[i] = max(sell[i - 2] - prices[i], buy[i - 1]);
            sell[i] = max(buy[i - 1] + prices[i], sell[i - 1]);

        return sell[n - 1]

"""
create two array, buy and sell. buy[i] means we buy a stock at day i , and sell[i] means we sell a stock at day i.
"""


"""
Runtime: 40 ms, faster than 63.84% of Python3 online submissions for Best Time to Buy and Sell Stock with Cooldown.
Memory Usage: 14.4 MB, less than 5.62% of Python3 online submissions for Best Time to Buy and Sell Stock with Cooldown.
"""